pipeline_dp/private_beam.py

```diff
@@ -308,3 +315,117 @@ def __init__(self, fn: Callable, label: Optional[str] = None):
     def expand(self, pcol: pvalue.PCollection):
         return pcol | "flatten values" >> beam.ParDo(
             FlatMap._FlattenValues(map_fn=self._fn))
+
+
+class PrivateCombineFn(beam.CombineFn):
+
+    @abc.abstractmethod
+    def add_private_input(self, accumulator, input):
+        pass
+
+    @abc.abstractmethod
+    def extract_private_output(self, accumulator):
+        pass
+
+    @abc.abstractmethod
+    def request_budget(self,
+                       budget_accountant: budget_accounting.BudgetAccountant):
+        """Requests the budget.
+
+        It is called by PipelineDP during the construction of the computations.
+        The custom combiner can request a DP budget by calling
+        'budget_accountant.request_budget()'. The budget object needs to be
+        stored in a field of 'self'. It will be serialized and distributed
+        to the workers together with 'self'.
+
+        Warning: do not store 'budget_accountant' in 'self'. It is assumed to
+        live in the driver process.
+        """
+        pass
+
+    @abc.abstractmethod
+    def set_aggregate_params(self,
+                             aggregate_params: pipeline_dp.AggregateParams):
+        """Sets aggregate parameters
+
+        The custom combiner can optionally use it for own DP parameter
```

Comment:
Just a few clarifications:

1) I'm wondering about "optionally". Can we think of a case where data in AggregateParams (e.g. sensitivity params) will *not* be used?

2) Why aren't we just storing aggregate_params here instead of letting clients to override this method? Is that because the clients may want to store only *some* parameters? In this case we could possibly clarify:

"Use this to store the parameters that are needed to ensure differential privacy for your custom combiner (e.g. sensitivity parameters)"